File: 2021/day_09.py
#!/usr/bin/python3

# read input as 2d list of numbers
input = open( "day_09.input")
data = [ list(map( int, x.strip())) for x in input.readlines() ]
input.close()

# helper functions
neighborhood = [
    [ -1,  0], # U
    [  1,  0], # D
    [  0, -1], # L
    [  0,  1]  # R
]

def getNeighbors( row, col, grid):
    maxRow = len(grid)
    maxCol = len(grid[0])
    neigh = []
    for r, c in neighborhood:
        r += row
        c += col
        if r >= 0 and c >= 0 and r < maxRow and c < maxCol:
            neigh.append({
                "val": grid[r][c],
                "row": r,
                "col": c
            })
    return neigh

def getBasinSize( row, col, heightmap):
    heightmap[row][col] = 9
    queue = [[ row, col]]
    size = 1
    for r, c in queue:
        neigh = getNeighbors( r, c, heightmap)
        for n in neigh:
            if n["val"] != 9:
                heightmap[n["row"]][n["col"]] = 9
                queue.append([ n["row"], n["col"]])
                size += 1
    return size

# solutions
def silver():
    heightmap = list(filter( lambda row: len(row) > 0, data))
    lowPoints = 0
    for r, row in enumerate(heightmap):
        for c, val in enumerate(row):
            neigh = getNeighbors( r, c, heightmap)
            if len(list(filter( lambda n: n["val"] <= val, neigh))) == 0:
                lowPoints += val + 1
    return lowPoints

def gold():
    heightmap = list(filter( lambda row: len(row) > 0, data))
    basins = []
    for r, row in enumerate(heightmap):
        for c, val in enumerate(row):
            if val != 9:
                basins.append(getBasinSize( r, c, heightmap))
    basins.sort()
    basins.reverse()
    return basins[0] * basins[1] * basins[2]

File: 2021/day_09.js
/* --- Day 9: Smoke Basin --- */

const neighborhood = [
    [ -1,  0], // U
    [  1,  0], // D
    [  0, -1], // L
    [  0,  1]  // R
];

function getNeighbors( row, col, grid){
    return neighborhood
        .map( ([ r, c]) => [ r + row, c + col])
        .filter( ([ r, c]) => r >= 0 && c >= 0 && r < grid.length && c < grid[0].length)
        .map( ([ r, c]) => ({
            val: grid[r][c],
            row: r,
            col: c
        }));
}

function getBasinSize( row, col, heightmap){
    let queue = [[ row, col]],
        size = 1;
    heightmap[row][col] = 9;
    while( queue.length ){
        let [ row, col] = queue.shift();
        getNeighbors( row, col, heightmap).forEach( n => {
            if( n.val != 9 ){
                queue.push([ n.row, n.col]);
                size++;
                heightmap[n.row][n.col] = 9;
            }
        });
    }
    return size;
}

function day_09a(){
    let heightmap = parseInput( x => x.trim().split("").map(Number)),
        lowPoints = 0;
    for( let row = 0; row < heightmap.length; row++ ){
        for( let col = 0; col < heightmap[0].length; col++ ){
            let val = heightmap[row][col];
            if( getNeighbors( row, col, heightmap).every( n => n.val > val) ){
                lowPoints += val + 1;
            }
        }
    }
    return lowPoints;
}

function day_09b(){
    let heightmap = parseInput( x => x.trim().split("").map(Number)),
        basins = [];
    for( let row = 0; row < heightmap.length; row++ ){
        for( let col = 0; col < heightmap.length; col++ ){
            if( heightmap[row][col] != 9 ){
                basins.push(getBasinSize( row, col, heightmap));
            }
        }
    }
    return prod(basins.sort(numericSortDesc).slice(0,3));
}
